source/common/matcher/prefix_map_matcher.h

```diff
@@ -30,17 +31,28 @@ template <class DataType> class PrefixMapMatcher : public MapMatcher<DataType> {
                    absl::optional<OnMatch<DataType>> on_no_match, absl::Status& creation_status)
       : MapMatcher<DataType>(std::move(data_input), std::move(on_no_match), creation_status) {}
 
-  absl::optional<OnMatch<DataType>> doMatch(const std::string& data) override {
-    const auto result = children_.findLongestPrefix(data);
-    if (result) {
-      return *result;
+  absl::optional<OnMatch<DataType>> doMatch(absl::string_view key, const DataType& data) override {
+    const absl::InlinedVector<std::shared_ptr<OnMatch<DataType>>, 4> results =
+        children_.findMatchingPrefixes(key);
+    bool retry_shorter = Runtime::runtimeFeatureEnabled(
+        "envoy.reloadable_features.prefix_map_matcher_resume_after_subtree_miss");
+    for (auto it = results.rbegin(); it != results.rend(); ++it) {
+      const auto& result = *it;
+      if (result->action_cb_ || !retry_shorter) {
+        return *result;
+      }
+      ASSERT(result->matcher_);
+      typename MatchTree<DataType>::MatchResult match = result->matcher_->match(data);
+      if (match.match_state_ != MatchState::MatchComplete || match.on_match_.has_value()) {
+        return match.on_match_;
+      }
     }
-
     return absl::nullopt;
   }
 
 private:
   TrieLookupTable<std::shared_ptr<OnMatch<DataType>>> children_;
+  bool retry_shorter_;
```

Comment:
Delete or initialize and remove the lookup above.

Reply:
Done.